walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java

```diff
@@ -270,5 +278,378 @@ public void notEnoughRoom() {
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test 
+  public void threeOptionalChallengeQuery() {
+    // A challenge query that involves three optional attendees 
+    // but only two are combatible with req attendee open slots.
+    //
+    //           |--------------D--------------|
+    //                       |----C----|   
+    // Opt.    : |--B--|           |--B--|
+    // Events  :       |--A--|     
+    // Day     : |-----------------------------|
+    // Options :                         |--2--|
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.fromStartDuration(TIME_0930AM, DURATION_60_MINUTES),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 4", TimeRange.fromStartDuration(TIME_0830AM, DURATION_90_MINUTES),
+            Arrays.asList(PERSON_C)),
+        new Event("Event 5", TimeRange.WHOLE_DAY, Arrays.asList(PERSON_D)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A), DURATION_30_MINUTES);
+
+    request.addOptionalAttendee(PERSON_B);
+    request.addOptionalAttendee(PERSON_C);
+    request.addOptionalAttendee(PERSON_D);
+
+    Collection<TimeRange> actual = query.optimalQuery(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TIME_1030AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test 
+  public void optimizedOptionalAttendeesBooked() {
+    // Have each person have different events. We should see three options because each person has
+    // split the restricted times. The optional attendee is fully-booked, so they have no effect
+    // on the result.
+    //
+    //           |--------------A--------------|
+    // Opt.    : |--------------B--------------|
+    // Events  :       
+    // Day     : |-----------------------------|
+    // Options : 
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.WHOLE_DAY,
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.WHOLE_DAY,
+            Arrays.asList(PERSON_B)));
+
+    MeetingRequest request =
+        new MeetingRequest(NO_ATTENDEES, DURATION_30_MINUTES);
+
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.optimalQuery(events, request);
+    Collection<TimeRange> expected = Arrays.asList();
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test 
+  public void optimizedDurationTest() {
```

Comment:
I recommend naming tests as predicates, which is to say sentence fragments that could be preceded with "This test verifies that the class…".  So this one might be "returnsLargestOpenDurationForOptimizedQuery".